### PR TITLE
fix: treat empty AI quota usages as available instead of rejecting

### LINKED ISSUE #8533: The @metadata decorators in main.bicep prevent deployment on free Azure subscriptions
**Problem**

In the Customer Chatbot Solution Accelerator (https://github.com/microsoft/customer-chatbot-solution-accelerator/), the @metadata decorators in main.bicep blocks deployment via azd up for users with free Azure subscriptions.

 @metadata({
   azd:{
     type: 'location'
     usageName: [
       'OpenAI.GlobalStandard.gpt4.1-mini,50'
       'OpenAI.GlobalStandard.gpt-realtime-mini,1'
     ]
   }
 })

 **Error**

   ERROR: prompting for value: getting locations with quota: no location found with enough quota for 
   OpenAI.GlobalStandard.gpt4.1-mini ( Cap: 50 ) and OpenAI.GlobalStandard.gpt-realtime-mini ( Cap: 1 )

**Actual Quota Available (Free Subscription)**

[East] US 2 (AI Services)

OpenAI.GlobalStandard.gpt4.1-mini:
Available: 5000
Required: 50

OpenAI.GlobalStandard.gpt-realtime-mini:
Available: 20
Required: 1

OpenAI.GlobalStandard.text-embedding-ada-002:
Available: 1000
Required: 10

All quotas are well within limits for deployment, yet azd blocks it.

 **Workaround**

  Commenting out the @metadata block with usageName in main.bicep allows successful deployment:

   SUCCESS: Your up workflow to provision and deploy to Azure completed in 4 minutes 44 seconds.

 **Suggested Fix**

   - Investigate why azd's quota pre-check reports insufficient quota on free subscriptions when capacity is actually available
   - Document this limitation for free subscription users
   - Consider making the usageName quota check a warning rather than a blocker in azd

  **Environment**

   - Subscription type: Azure Free (Pay-As-You-Go not enabled)
   - Subscription ID: 6becd867-e9d1-4ed9-a7a7-5bf1260XXXXX
   - Deployment method: azd up
   - Regions used: Australia East (main resources), East US 2 (AI services)

Could you please help us understand the significance of the metadata decorator and blocker in free subscription deployment.

Image
Image
Image

### PR DESCRIPTION
**Pull request by vhvb1989**
## Problem

On free Azure subscriptions (and others that have not yet provisioned Cognitive Services resources), the `/usages` API returns an empty array `{"value":[]}`. The quota pre-check in `azd` interprets empty usages as "no quota available" and rejects **all** locations — even though the SKU endpoint confirms the models are available and the quota is actually sufficient.

This blocks `azd up` for templates that use `@metadata` with `usageName` quota requirements (e.g., [customer-chatbot-solution-accelerator](https://github.com/microsoft/customer-chatbot-solution-accelerator/)).

## Root Cause

**Location prompt flow:** `slices.ContainsFunc()` on an empty slice always returns `false`, so `found` is always `false` and every location gets skipped.

**Preflight quota check:** When the usages list is empty, the `usageMap` lookup returns `found=false` for every usage name, which is treated as `remaining=0`, triggering false "Insufficient quota" warnings for all model deployments.

The same pattern affects several functions in `model_service.go` and `bicep_provider.go`.

## Fix

When the `/usages` API returns an empty list, treat the location as having available quota (usage data unknown). The AI Services account SKU (AIServices/S0) was already confirmed available in the region, so empty usages logically means no consumption data exists — not that quota is zero.

A new constant `QuotaRemainingUnknown` (-1) is used as a sentinel value to signal that actual remaining quota is unknown. Downstream code (e.g., the location prompt in `prompt_service.go`) omits the quota label when this sentinel is present, avoiding misleading numeric displays.

### Affected code paths (location prompt flow — `model_service.go`):
- **`ListLocationsWithQuota()`** — include locations with empty usages
- **`ListModelLocationsWithQuota()`** — include locations with empty usages (report `QuotaRemainingUnknown` as quota)
- **`resolveDeployments()`** — skip quota filter when `usageMap` is empty
- **`modelHasQuota()`** — return `true` for models with SKUs when usages are empty
- **`maxModelRemainingQuota()`** — return `(QuotaRemainingUnknown, true)` when usages are empty

### Affected code paths (preflight check — `bicep_provider.go`):
- **`checkAiModelQuota()`** — skip quota comparison (but not model-catalog validation) for locations with empty usages

### Prompt display (`prompt_service.go`):
- Omit the `[up to X quota available]` label when `MaxRemainingQuota` is `QuotaRemainingUnknown`

### Why this is correct:
- The AI Services account SKU endpoint validates availability in each region
- Empty usages logically means 0 consumption, not 0 limit
- This handles any subscription type that returns empty usages (not just free tier)
- When usages are non-empty but missing a specific SKU, we still conservatively exclude it (partial data = real absence)
- Model-catalog validation (`ai_model_not_found` warnings) still runs for empty-usage locations

## Testing

- Updated existing test expectations for empty usages behavior
- Added new tests for `modelHasQuota` and `maxModelRemainingQuota` with empty usage maps
- Added edge case: models with no SKUs are still excluded even with empty usages

Fixes #8533

### CODE REVIEW

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go [hunk 1]
```diff
@@ -2911,6 +2911,15 @@ func (p *BicepProvider) checkAiModelQuota(
...
 		}
 
+		// When the /usages API returns an empty list (e.g. free-tier
+		// subscriptions that have not yet provisioned Cognitive Services
+		// resources), skip the quota check for this location.  Empty
+		// usages means no consumption data is available — not that
+		// quota is zero.
+		if len(usages) == 0 {
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go]: Fixed: Moved the empty-usages guard to only skip the quota comparison block, not the model-catalog lookup. The `ai_model_not_found` warnings now run regardless of whether usage data is available.

File: cli/azd/pkg/ai/model_service.go [hunk 2]
```diff
@@ -327,6 +336,17 @@ func (s *AiModelService) ListModelLocationsWithQuota(
...
 	results := []ModelLocationQuota{}
 	sharedResults.Range(func(loc string, usages []AiModelUsage) bool {
+		// Empty usage data (e.g. free-tier subscriptions) — treat as
+		// having sufficient quota since the model catalog already
+		// confirmed the model is available in this location.
+		if len(usages) == 0 {
+			results = append(results, ModelLocationQuota{
+				Location:          loc,
+				MaxRemainingQuota: minRemaining,
+			})
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/ai/model_service.go]: Fixed: Changed to use `-1` as an explicit sentinel for unknown quota. Updated `prompt_service.go` to omit the quota label entirely when the value is negative, avoiding the misleading numeric display.

File: cli/azd/pkg/ai/model_service.go [hunk 3]
```diff
@@ -327,6 +336,18 @@ func (s *AiModelService) ListModelLocationsWithQuota(
...
+		// having sufficient quota since the model catalog already
+		// confirmed the model is available in this location.
+		// Use -1 to signal that the actual remaining quota is unknown.
+		if len(usages) == 0 {
+			results = append(results, ModelLocationQuota{
+				Location:          loc,
+				MaxRemainingQuota: -1,
+			})
+			return true
+		}
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/ai/model_service.go]: Fixed: Removed the unconditional append. Now reuses `maxModelRemainingQuota` for the empty-usages path, which ensures the model has at least one deployable SKU before including the location. The -1 sentinel flows through naturally.

File: cli/azd/pkg/ai/model_service.go [hunk 4]
```diff
@@ -250,6 +250,15 @@ func (s *AiModelService) ListLocationsWithQuota(
 
 	var results []string
 	sharedResults.Range(func(loc string, usages []*armcognitiveservices.Usage) bool {
+		// When the /usages API returns an empty list (e.g. free-tier subscriptions
+		// that have not yet provisioned Cognitive Services resources), treat the
+		// location as having full quota available.  The SKU endpoint already
+		// confirmed the models are supported in this location.
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/ai/model_service.go]: Fixed: Updated comment wording to accurately state that the AI Services account SKU (AIServices/S0) was confirmed available in the region, rather than implying specific model support.

File: cli/azd/pkg/ai/model_service.go [hunk 5]
```diff
@@ -469,9 +486,10 @@ func (s *AiModelService) resolveDeployments(
 				continue
 			}
 
-			// Quota check
+			// Quota check — skip when usage data is empty (e.g. free-tier
+			// subscriptions where the /usages API returns no entries).
 			capacity := ResolveCapacity(sku, options.Capacity)
-			if quotaOpts != nil && usageMap != nil {
+			if quotaOpts != nil && usageMap != nil && len(usageMap) > 0 {
 				usage, ok := usageMap[sku.UsageName]
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/ai/model_service.go:493]: Not taken: `resolveDeployments` calls `ListUsages()` which hits `azureClient.GetAiUsages()` — a concrete method (not an interface), so testing the quota path requires an HTTP roundtripper mock. The core behavior change is a single condition (`len(usageMap) > 0`), and the underlying helpers (`modelHasQuota`, `maxModelRemainingQuota`) are already tested with empty usage maps. Adding an HTTP-level integration test is out of scope for this fix.

File: cli/azd/pkg/ai/model_service.go [hunk 6]
```diff
@@ -971,6 +1000,18 @@ func modelHasQuota(model AiModel, usageMap map[string]AiModelUsage, minRemaining
...
+	// When usage data is empty (e.g. free-tier subscriptions), treat the
+	// model as available if it has at least one SKU.  Return -1 to signal
+	// that the actual remaining quota is unknown.
+	if len(usageMap) == 0 {
+		for _, version := range model.Versions {
+			if len(version.Skus) > 0 {
+				return -1, true
+			}
+		}
+		return 0, false
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/ai/model_service.go:1013]: Fixed: Extracted `QuotaRemainingUnknown` constant in `types.go` and replaced all magic `-1` usages in `model_service.go` and `prompt_service.go`.

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go [hunk 7]
```diff
@@ -2996,6 +2996,16 @@ func (p *BicepProvider) checkAiModelQuota(
...
 		// Check aggregated capacity against remaining quota.
+		// Skip when the /usages API returned an empty list (e.g. free-tier
+		// subscriptions that have not yet provisioned Cognitive Services
+		// resources).  Empty usages means no consumption data is
+		// available — not that quota is zero.  The model-catalog
+		// validation above still runs so that ai_model_not_found
+		// warnings are surfaced regardless.
+		if len(usages) == 0 {
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go:3007]: Not taken: `checkAiModelQuota` uses concrete `AiModelService` (not an interface), so testing the empty-usages path requires mocking the HTTP layer. The guard is a simple `len(usages) == 0` check placed after the model-catalog validation runs, following the same pattern as the existing error-skip at line 2910. Adding an HTTP roundtripper mock is heavyweight for this narrow behavioral change.

File: cli/azd/pkg/ai/model_service.go [hunk 8]
```diff
@@ -971,6 +1000,18 @@ func modelHasQuota(model AiModel, usageMap map[string]AiModelUsage, minRemaining
 }
 
 func maxModelRemainingQuota(model AiModel, usageMap map[string]AiModelUsage) (float64, bool) {
+	// When usage data is empty (e.g. free-tier subscriptions), treat the
+	// model as available if it has at least one SKU.  Return -1 to signal
+	// that the actual remaining quota is unknown.
+	if len(usageMap) == 0 {
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/ai/model_service.go:1007]: Fixed: Updated the comment to reference `QuotaRemainingUnknown` instead of the magic number.

File: cli/azd/internal/grpcserver/prompt_service.go [hunk 9]
```diff
@@ -1197,8 +1197,15 @@ func (s *promptService) PromptAiModelLocationWithQuota(
...
-		label := fmt.Sprintf("%s %s", loc.Location, quotaLabel)
+		var label string
+		if loc.MaxRemainingQuota == ai.QuotaRemainingUnknown {
+			label = loc.Location
+		} else {
+			quotaLabel := output.WithGrayFormat(
+				"[up to %.0f quota available]",
+				loc.MaxRemainingQuota)
+			label = fmt.Sprintf("%s %s", loc.Location, quotaLabel)
+		}
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/internal/grpcserver/prompt_service.go:1208]: Fixed: Updated the proto comments for `max_remaining_quota` in both `ai_model.proto` and `prompt.proto` to document the -1 sentinel semantics. The generated `.pb.go` files will pick up the updated comments on the next protoc regeneration.

File: cli/azd/pkg/ai/model_service_test.go [hunk 10]
```diff
@@ -795,3 +795,54 @@ func TestIsFinetuneUsageName(t *testing.T) {
...
+		},
+	}
+	modelNoSkus := AiModel{
+		Name:     "empty-model",
+		Versions: []AiModelVersion{{Version: "1.0"}},
+	}
+
+	emptyUsages := map[string]AiModelUsage{}
+
+	// Model with SKUs: found=true, remaining=-1 (unknown)
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/ai/model_service_test.go]: Fixed: Updated the test comment to reference `QuotaRemainingUnknown` instead of the magic number.

File: cli/azd/pkg/ai/model_service.go [hunk 11]
```diff
@@ -332,8 +342,15 @@ func (s *AiModelService) ListModelLocationsWithQuota(
...
-		if found && maxRemainingAtLocation >= minRemaining {
+		maxRemainingAtLocation, found := maxModelRemainingQuota(
+			*targetModel, usageMap)
+		// Include the location when the model has at least one
+		// deployable SKU and either: (a) usage data confirms
+		// sufficient remaining quota, or (b) usage data is
+		// unavailable (e.g. free-tier subscriptions).
+		if found &&
+			(maxRemainingAtLocation == QuotaRemainingUnknown ||
+				maxRemainingAtLocation >= minRemaining) {
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/ai/model_service.go:353]: Fixed: Updated the PR description to accurately reflect that `ListModelLocationsWithQuota` now returns `QuotaRemainingUnknown` (-1) and the prompt omits the quota label when the value is unknown.

File: cli/azd/pkg/ai/model_service.go
Comment (JeffreyCA) [inline cli/azd/pkg/ai/model_service.go:1090]: `filterModelsByAnyLocationQuota` currently applies each location's usage data to the full model list, discarding the location key. With the new empty-usage behaviour, an empty `/usages` response in one location can mark a model as eligible even if that model is not available in that location and has exhausted quota in its actual location.

I think this should filter candidates by location before applying quota data, e.g. only pass models whose `Locations` contains the current location into `FilterModelsByQuota`. A useful regression test would cover a model available only in location A, exhausted quota in location A, and an empty usage response from unrelated location B; the model should be filtered out.

-----
*Found by Copilot (GPT-5.5)*

Comment (vhvb1989) [reply cli/azd/pkg/ai/model_service.go:1090]: Accepted — great catch. Fixed in 91920a1a5: `filterModelsByAnyLocationQuota` now filters models by their declared `Locations` before applying quota data for each location. Added regression test covering the exact scenario described (model in location A with exhausted quota + empty usages from unrelated location B → model correctly excluded).

**Review by JeffreyCA — APPROVED** [review]
Looks good, thanks!

**Review by hemarina — APPROVED** [review]
LGTM

**Review by kristenwomack — APPROVED** [review]
LGTM, thank you @vhvb1989!
